Added example for :class:`~.Sphere`

jojo96 opened the pull request:
<!--
Thanks for your contribution to ManimCommunity!

Before filling in the details, ensure:
- Your local changes are up-to-date with ManimCommunity/manim
  
- The title of your PR gives a descriptive summary to end-users. Some examples:
  - Fixed last animations not running to completion
  - Added gradient support and documentation for SVG files
  Examples of what *NOT* to do:
  - "fixed that styling issue" - not descriptive enough
  - "fixed issue #XYZ" - end-user needs to do further research
-->
## Changelog / Overview
<!-- Optional (Recommended): a detailed overview of the PR for the upcoming
release's changelog entry. Useful for when the PR title isn't enough. 

DO NOT REMOVE THE FOLLOWING CHANGELOG LINES, EVEN IF YOU DON'T USE THEM.-->
<!--changelog-start-->
Added documentation for the sphere class. Added an example since there were no existing examples for the sphere class
<!--changelog-end-->

## Motivation
<!-- In what way do your changes improve the library? -->
Added example for new users
## Explanation for Changes
<!-- How do your changes improve the library? -->
Makes it more readable as the examples are useful to the newer users
<!-- For PRs introducing new features, please adjust
the link below to reference the docs of your feature. -->
[Documentation Reference](https://manimce--1627.org.readthedocs.build/en/1627/reference/manim.mobject.three_dimensions.Sphere.html)
<!-- In the link above replace #### with you PR number.
You can also adjust the path to the module / class you worked on. This could look like adding
"/en/####/reference/manim.mobject.geometry.Arc.html" to the link for the Arc class.
Notice that the link will only work once the documentation is build which may take a few minutes.-->


## Testing Status
<!-- Optional (Recommended): your computer specs and what tests you ran with
their results, if any. This section is also intended for other
testing-related comments. -->

## Further Comments
<!-- Optional: any further comments that might be useful for reviewers. -->

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [ ] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings
<!-- Once again, thanks for contributing to ManimCommunity! -->


<!-- Do not modify the lines below. These are for the reviewers of your PR -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [x] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

Darylgolden commented:
[Courtesy link](https://manimce--1627.org.readthedocs.build/en/1627/reference/manim.mobject.three_dimensions.Sphere.html?highlight=sphere#manim.mobject.three_dimensions.Sphere)

Darylgolden commented:
Hi @jojo96, thank you for your contribution! Personally I don't like the look of the spheres overlapping, could you change that perhaps? It would be nice to have different colors as well.

kolibril13 commented:
Thanks for the contribution!
Two suggestions:
A) I also think it is better when the spheres do not overlap.
B) I would name them sphere1, sphere2, sphere3

jojo96 commented:
> Hi @jojo96, thank you for your contribution! Personally I don't like the look of the spheres overlapping, could you change that perhaps? It would be nice to have different colors as well.


> Thanks for the contribution!
> Two suggestions:
> A) I also think it is better when the spheres do not overlap.
> B) I would name them sphere1, sphere2, sphere3

Hi I have created spheres with better resolutions and changed colors as well. Also, implemented A and B as suggested. Added these to patch-2. 

Do I need to create a new pull request? @kolibril13

kolibril13 commented:
Great! 
No, you don't have to recreate a pr.
You just have to wait until this test is passing,
![image](https://user-images.githubusercontent.com/44469195/120895424-2d391900-c61d-11eb-8f19-afcfafdcf89d.png)
and the new example will be available here:
https://manimce--1627.org.readthedocs.build/en/1627/reference/manim.mobject.three_dimensions.Sphere.html?highlight=sphere#manim.mobject.three_dimensions.Sphere

hydrobeam commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,30 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
+                sphere1 = Sphere(center = (3,0,0),radius = 1,resolution = (30, 40),u_min = 0.001,u_max = PI - 0.001,v_min = 0,v_max = TAU)
+                sphere1.set_color(RED)
+                self.add(sphere1)
+                sphere2 = Sphere(center=(0,1,0),radius=2,resolution=(40, 50))
```
What's the reasoning behind specifying `center` and then using `move_to` ? (same goes for sphere3)

jojo96 replied:
Removed the move_to functions.

hydrobeam commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,30 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
+                sphere2.move_to((-1,-3,0))
+                sphere2.set_color(GREEN)
+                self.add(sphere2)
+                sphere3 = Sphere(center=(0,1,0),radius=2,resolution=(40, 50))
```
All these examples are pretty high res. I would recommend showing what a low res sphere would look like too. This is a sample value I toyed around with. 
```suggestion
                sphere3 = Sphere(center=(0,1,0),radius=2,resolution=(14, 14))
```

hydrobeam replied:
This is a bit of an afterthought but: dropping the resolution of sphere1 to `(20, 30)` and sphere2 to `(35, 40)` would hardly affect their look and save a couple seconds in rendering time.

jojo96 replied:
@naveen521kk removed sphere from top line. Also copied the documentation from the bot. 

And also there is no hackathon label added to this thread.

hydrobeam reviewed (changes requested):
Styling + more variety in resolutions

naveen521kk commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,30 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
 
 
 class Sphere(ParametricSurface):
+    """A sphere
```
I think this docstring too obvious. Either add few more lines or remove this.

behackl replied:
> I think this docstring too obvious. Either add few more lines or remove this.

No, that's unfortunately bad advice. Every docstring **needs** to start with a one-line summary. If it is just "A Sphere.", then that's also just fine. A longer version ("A Mobject representing a threedimensional sphere.") would also be fine.

naveen521kk commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,30 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
+                sphere3.move_to((-1,2,0))
+                sphere3.set_color(BLUE)
+                self.add(sphere3)
+
```
FYI, using the bot should do this automatically.

behackl commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,28 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
 
 
 class Sphere(ParametricSurface):
+    """
```
Just as mentioned in the other comment: every docstring **needs** to start with a one-line summary. Please add back something along the lines of

```suggestion
    """A mobject representing a three-dimensional sphere.
```

(feel free to commit your own suggestion instead.)

behackl reviewed (changes requested):
Left one change request. @hydrobeam's comment regarding the resolution (which has a rather large impact on the render time in our documentation) also seems to have not been addressed, could you please take a look again?

Thank you!

jojo96 commented:
> Left one change request. @hydrobeam's comment regarding the resolution (which has a rather large impact on the render time in our documentation) also seems to have not been addressed, could you please take a look again?
> 
> Thank you!

@behackl reduced the resolution of the spheres to (20,20), (18,18), (16,16). Reducing the resolution further does not render good looking spheres.

behackl reviewed (approved):
I'm happy with this; slightly higher resolution would probably be fine as well. Let's see what @hydrobeam has to say.

hydrobeam commented on manim/mobject/three_dimensions.py:
```diff
@@ -134,6 +134,28 @@ def set_fill_by_checkerboard(self, *colors, opacity=None):
+    .. manim:: ExampleSphere
+        :save_last_frame:
+
+        class ExampleSphere(ThreeDScene):
+            def construct(self):
+                self.set_camera_orientation(phi=PI / 6, theta=PI / 6)
+                sphere1 = Sphere(center=(3, 0, 0), radius=1, resolution=(20, 20), u_min=0.001, u_max=PI - 0.001, v_min=0, v_max=TAU)
+                sphere1.set_color(RED)
+                self.add(sphere1)
+                sphere2 = Sphere(center=(-1, -3, 0), radius=2, resolution=(18, 18))
+                sphere2.set_color(GREEN)
+                self.add(sphere2)
+                sphere3 = Sphere(center=(-1, 2, 0), radius=2, resolution=(16, 16))
+                sphere3.set_color(BLUE)
+                self.add(sphere3)
```
```suggestion
        class ExampleSphere(ThreeDScene):
            def construct(self):
                self.set_camera_orientation(phi=PI / 6, theta=PI / 6)
                sphere1 = Sphere(
                    center=(3, 0, 0),
                    radius=1,
                    resolution=(20, 20),
                    u_min=0.001,
                    u_max=PI - 0.001,
                    v_min=0,
                    v_max=TAU,
                )
                sphere1.set_color(RED)
                self.add(sphere1)
                sphere2 = Sphere(center=(-1, -3, 0), radius=2, resolution=(18, 18))
                sphere2.set_color(GREEN)
                self.add(sphere2)
                sphere3 = Sphere(center=(-1, 2, 0), radius=2, resolution=(16, 16))
                sphere3.set_color(BLUE)
                self.add(sphere3)
```

hydrobeam reviewed (approved):
Looks good to me too. Just apply black formatting here so that you don't have to scroll horizontally to see all the code.

Darylgolden reviewed (approved):
LGTM
